Duration/Memory Convenience Click Types

TimothyWillard opened the pull request:
### Describe your changes.

This pull request add convenience click types for durations and memory. This is particularly useful in specifying time/memory limits for batch jobs in a user friendly way (i.e. `--time=30min` or `--memory=4GB`). While memory is probably specific to batch, durations could be used in other actions such as specifying forecast intervals.

### Does this pull request make any user interface changes? If so please describe.

The user interface changes are adding `DurationParamType` and `MemoryParamType` to `gempyor.shared_cli`, but these changes are new and have no affect on existing user interfaces.

Those are reflected in updates to the documentation in docstrings of added functionality.

### What does your pull request address? Tag relevant issues.

This pull request was spun out of GH-394, while it does not correspond to an exact issue it does help move GH-365 along by shrinking that PR's size.

pearsonca commented on flepimop/gempyor_pkg/tests/shared_cli/test_duration_param_type_class.py:
```diff
@@ -0,0 +1,40 @@
+
+
+@pytest.mark.parametrize("nonnegative", (True, False))
+@pytest.mark.parametrize("value", ("abc", "$12.34", "12..3", "12years", "12.a2"))
```
is forbidding "years" about that being conceptually wrong or simply a duration we don't want to contemplate? seems as a generic thing, its appropriate for the tool to potentially support years, so a bit weird to explicitly test that it *doesn't*.

TimothyWillard replied:
Two issues with years: 1) years do not have a constant length (2024 was 366 days but most years are 365 days), and 2) python's `timedelta` class which doesn't have a `years` argument: https://docs.python.org/3/library/datetime.html#datetime.timedelta.

I could however see an argument for adding micro/milliseconds on conceptual ground if you would like? I just excluded those since it was too short for most practical use cases.

pearsonca replied:
so this is a duration we explicitly don't want to contemplate (primarily because it's an ambiguous unit for our case) - that's fine! It does feel worth noting somewhere that we don't allow this unit of time - like maybe in the method documentation? But I don't think we need, say, a special error message for years.

re ms - I think it's fine to not support it _at this time_. I think you're saying this would be one that we'd be willing to contemplate (i.e. we aren't testing to ensure its exclusion), but that we haven't bothered to yet support (because we don't think its practically useful).

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/shared_cli.py:
```diff
@@ -124,6 +127,191 @@ def cli(ctx: click.Context) -> None:
+    Attributes:
+        name: The name of the parameter type.
+
+    Examples:
+        >>> from gempyor.shared_cli import DurationParamType
+        >>> duration_param_type = DurationParamType(False)
+        >>> duration_param_type.convert("23min", None, None)
+        datetime.timedelta(seconds=1380)
+        >>> duration_param_type.convert("2.5hr", None, None)
+        datetime.timedelta(seconds=9000)
+        >>> duration_param_type.convert("-2", None, None)
+        datetime.timedelta(days=-1, seconds=86280)
+    """
```
copy paste error?

TimothyWillard replied:
No, why?

pearsonca replied:
this is on the memoryparamtype - example documentation for the duration param type?

TimothyWillard replied:
Oops, thanks

pearsonca reviewed (changes requested):
Seems mostly fine. One question about forbidding years (seems like the type should conceptually support it, even if we don't actually yet or don't want to accept values that large).

Also feels like shared_cli might be getting a bit bloated (even before adding these).

Does it make sense to start having a click extensions submodule or somesuch? And then shared_cli is importing generic capabilities from there?

TimothyWillard commented:
> Also feels like shared_cli might be getting a bit bloated (even before adding these).

Do you want me to move these to `_click`?

pearsonca commented:
> > Also feels like shared_cli might be getting a bit bloated (even before adding these).
> 
> Do you want me to move these to `_click`?

If that's the standard approach to these kind of external-dependency extension/customizations, that seems the best way to manage it.

TimothyWillard commented:
> Docs question: where are we capturing how bare numbers behave for these types? As in, where would I look, other than code, to see if --time=5 is 5 seconds vs 5 minutes?

I'll document.

TimothyWillard commented:
- Extracted these two custom types from `gempyor.shared_cli` to `gempyor._click`,
- Added extra behavior around unitless values for both custom types including a way to outright reject said values if wanted along with corresponding docs/tests, and
- Corrected documentation typos.
